code/Serving_API.py

```diff
@@ -1,27 +1,258 @@
+        coordCityMap[end_city_coord] = to_city_info
+        coord_set.append(end_city_coord)
+
+        edge_type = None
+
+        if option == 'submarine_standard_paths':
+            edge_type = "submarine"
+        else:
+            edge_type = 'land'
+
+        add_edge(G, from_city_info,
+                    to_city_info, distance_km, path_wkt, start_city_coord, end_city_coord, edge_type)
+        add_edge(G, to_city_info,
+                    from_city_info, distance_km, reverse_linestring_order(path_wkt), end_city_coord, start_city_coord, edge_type)
+    return G, coordCityMap, coord_set
+
+
+def build_up_global_graph(db_file):
+    submarine_standard_paths = get_all_submarine_standard_paths(db_file)
+
+    submarine_to_standard_paths_pairs = get_all_submarine_to_standard_paths_pairs(
+        db_file)
+    standard_paths = get_all_standard_paths(db_file)
+    G = nx.DiGraph()
+    coordCityMap = {}
+    coord_set = []
+
+    G, coordCityMap, coord_set = graph_build_helper(
+        G, coordCityMap, coord_set, standard_paths, 'standard_paths')
+    G, coordCityMap, coord_set = graph_build_helper(
+        G, coordCityMap, coord_set, submarine_standard_paths, 'submarine_standard_paths')
+    G, coordCityMap, coord_set = graph_build_helper(
+        G, coordCityMap, coord_set, submarine_to_standard_paths_pairs, 'submarine_to_standard_paths_pairs')
+
+    return coordCityMap, coord_set, G
+
 
 app = FastAPI()
 
+
 @app.get("/physical-route/")
 def physical_route(src_latitude: float, src_longitude: float,
                    dst_latitude: float, dst_longitude: float) -> list[tuple[float, float]]:
-    """Get the physical route in (lat, lon) format from src to dst, including both ends.
-
-    Returns:
-        A list of (lat, lon) tuples, one for each hop along the route. This should include both ends.
     """
-    # TODO: get the real physical routes from database.
-    return [
-        (src_latitude, src_longitude),
-        (0, 0),
-        (dst_latitude, dst_longitude),
-    ]
+    Get the physical route in (lat, lon) format from src to dst, including both ends.
+    """
+    # Convert input coordinates to city information
+    src_city_co = find_closest_point(
+        (src_latitude, src_longitude), coord_set)
+
+    dst_city_co = find_closest_point(
+        (dst_latitude, dst_longitude), coord_set)
+
+    src_city_info = coordCityMap[src_city_co]
+    dst_city_info = coordCityMap[dst_city_co]
+
+    assert src_city_info and dst_city_info
+
+    # Find shortest path between cities in the graph
+    try:
+        shortest_path_cities = nx.shortest_path(
+            G, source=src_city_info, target=dst_city_info, weight='distance')
+
+        shortest_distance, coordinate_list, wkt_list, cable_type_list = calculate_shortest_path_distance(
+            G, shortest_path_cities)
+
+        return (coordinate_list, wkt_list, cable_type_list)
+    except nx.NetworkXNoPath:
+        print("cannot find the shortest path")
```

Comment:
Report a 400 error. See example [here](https://fastapi.tiangolo.com/tutorial/handling-errors/#import-httpexception)

Reply:
raised a 400 Bad request error when there is no shortest path between two cities.